agent: Allow customizing temperature by provider/model

agu-z opened the pull request:
Adds a new `agent.model_parameters` setting that allows the user to specify a custom temperature for a provider AND/OR model:

```json5
    "model_parameters": [
      // To set parameters for all requests to OpenAI models:
      {
        "provider": "openai",
        "temperature": 0.5
      },
      // To set parameters for all requests in general:
      {
        "temperature": 0
      },
      // To set parameters for a specific provider and model:
      {
        "provider": "zed.dev",
        "model": "claude-3-7-sonnet-latest",
        "temperature": 1.0
      }
    ],
```

Release Notes:

- agent: Allow customizing temperature by provider/model

agu-z committed:
Add agent.models setting

Co-authored-by: Max Brunsfeld <maxbrunsfeld@gmail.com>

agu-z committed:
Structure the model settings as an array of rules

Co-authored-by: Max Brunsfeld <maxbrunsfeld@gmail.com>

maxdeviant committed:
Format `default.json`

maxdeviant commented on crates/assistant_settings/src/assistant_settings.rs:
```diff
@@ -695,8 +784,6 @@ pub struct LegacyAssistantSettingsContent {
 impl Settings for AssistantSettings {
     const KEY: Option<&'static str> = Some("agent");
 
-    const FALLBACK_KEY: Option<&'static str> = Some("assistant");
```
@agu-z @maxbrunsfeld Was this removal intentional? Cole just added it a few hours ago in https://github.com/zed-industries/zed/pull/29943.

maxdeviant replied:
Actually, seems like a lot of that PR got clobbered by this one, like the tests further down in this file.

agu-z replied:
Alright, restored! That was really strange

agu-z committed:
Restore lost code

agu-z committed:
Fix test

agu-z committed:
Fix typo

agu-z committed:
Register AssistantSettings in assistant_context_editor tests

maxbrunsfeld committed:
Initialize agent settings in tests

agu-z committed:
More 📎 fixes

maxbrunsfeld committed:
Merge branch 'main' into model-parameters

maxbrunsfeld committed:
Fix spurious migration of openai + anthropic settings

Previously, even settings with the "version": "1" property
were treated as legacy.

Co-authored-by: Agus Zubiaga <hi@aguz.me>
